rpc/core/blocks.go

```diff
@@ -181,58 +181,64 @@ func (env *Environment) BlockSearch(
+		return nil, errors.New("block searching is disabled due to no kvEventSink")
 	}
 
 	q, err := tmquery.New(query)
 	if err != nil {
 		return nil, err
 	}
 
-	results, err := env.BlockIndexer.Search(ctx.Context(), q)
-	if err != nil {
-		return nil, err
-	}
-
-	// sort results (must be done before pagination)
-	switch orderBy {
-	case "desc", "":
-		sort.Slice(results, func(i, j int) bool { return results[i] > results[j] })
+	for _, sink := range env.EventSinks {
+		if sink.Type() == indexer.KV {
+			results, err := sink.SearchBlockEvents(ctx.Context(), q)
+			if err != nil {
+				return nil, err
+			}
 
-	case "asc":
-		sort.Slice(results, func(i, j int) bool { return results[i] < results[j] })
+			// sort results (must be done before pagination)
+			switch orderBy {
+			case "desc", "":
+				sort.Slice(results, func(i, j int) bool { return results[i] > results[j] })
 
-	default:
-		return nil, fmt.Errorf("%w: expected order_by to be either `asc` or `desc` or empty", ctypes.ErrInvalidRequest)
-	}
+			case "asc":
+				sort.Slice(results, func(i, j int) bool { return results[i] < results[j] })
 
-	// paginate results
-	totalCount := len(results)
-	perPage := env.validatePerPage(perPagePtr)
+			default:
+				return nil, fmt.Errorf("%w: expected order_by to be either `asc` or `desc` or empty", ctypes.ErrInvalidRequest)
+			}
 
-	page, err := validatePage(pagePtr, perPage, totalCount)
-	if err != nil {
-		return nil, err
-	}
+			// paginate results
+			totalCount := len(results)
+			perPage := env.validatePerPage(perPagePtr)
 
-	skipCount := validateSkipCount(page, perPage)
-	pageSize := tmmath.MinInt(perPage, totalCount-skipCount)
+			page, err := validatePage(pagePtr, perPage, totalCount)
+			if err != nil {
+				return nil, err
+			}
 
-	apiResults := make([]*ctypes.ResultBlock, 0, pageSize)
-	for i := skipCount; i < skipCount+pageSize; i++ {
-		block := env.BlockStore.LoadBlock(results[i])
-		if block != nil {
-			blockMeta := env.BlockStore.LoadBlockMeta(block.Height)
-			if blockMeta != nil {
-				apiResults = append(apiResults, &ctypes.ResultBlock{
-					Block:   block,
-					BlockID: blockMeta.BlockID,
-				})
+			skipCount := validateSkipCount(page, perPage)
+			pageSize := tmmath.MinInt(perPage, totalCount-skipCount)
+
+			apiResults := make([]*ctypes.ResultBlock, 0, pageSize)
+			for i := skipCount; i < skipCount+pageSize; i++ {
+				block := env.BlockStore.LoadBlock(results[i])
+				if block != nil {
+					blockMeta := env.BlockStore.LoadBlockMeta(block.Height)
+					if blockMeta != nil {
+						apiResults = append(apiResults, &ctypes.ResultBlock{
+							Block:   block,
+							BlockID: blockMeta.BlockID,
+						})
+					}
+				}
 			}
+
+			return &ctypes.ResultBlockSearch{Blocks: apiResults, TotalCount: totalCount}, nil
 		}
 	}
 
-	return &ctypes.ResultBlockSearch{Blocks: apiResults, TotalCount: totalCount}, nil
+	return nil, errors.New("block querying is disabled on this node due to the KV event sink being disabled;" +
```

Comment:
does this get executed if you exit early due to not using the kvEventSink

Reply:
right, so I just give an unexpected return error here?